### PR TITLE
OCPNODE-1656: oc image extract: Introduce IDMS as alternative source

### PR DESCRIPTION
**Pull request by ardaguclu**
This PR;

* adds foundational work to support IDMS(aka ImageDigestMirrorSet) as an alternative source
in addition to the ICSP. 
* introduces `--idms-file` in `oc image extract` depending on the foundational work above ^
* deprecates `--icsp-file` flag in `oc image extract` command to suggest `--idms-file` instead

### CODE REVIEW

**Comment by ardaguclu** [comment]
/hold
until approach is validated and pre-merge testing is done

File: pkg/cli/image/strategy/onerror.go [hunk 1]
```diff
@@ -162,3 +154,150 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
...
+// addSourceAsLastAlternate decides whether the original imageRef is first or the last element in the result
+func alternativeImageSourcesIDMS(imageRef reference.DockerImageReference, idmsList []apicfgv1.ImageDigestMirrorSet, addSourceAsLastAlternate bool) ([]reference.DockerImageReference, error) {
+	var imageSources []reference.DockerImageReference
+	klog.V(5).Infof("%v ImageReference added to potential ImageSourcePrefixes from ImageDigestMirrorSet", imageRef.AsRepository().AsV2())
+	if !addSourceAsLastAlternate {
+		imageSources = append(imageSources, imageRef.AsRepository().AsV2())
+	}
+	repo := imageRef.AsRepository().Exact()
+	for _, idms := range idmsList {
+		repoDigestMirrors := idms.Spec.ImageDigestMirrors
```
Comment (ardaguclu) [inline pkg/cli/image/strategy/onerror.go:229]: This PR assumes that ICSP and IDMS have same functionality. If there is a behavioral difference between these two types, we need to make changes in here. For example, IDMS has `mirrorSourcePolicy` field and we ignores this field in here, is that correct @QiWang19 ?.

File: pkg/cli/image/strategy/onerror.go [hunk 2]
```diff
@@ -162,3 +154,150 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
...
+	imageRefList, err := alternativeImageSourcesIDMS(imageRef, idmsList, false)
+	if err != nil {
+		return nil, err
+	}
+	return imageRefList, nil
+}
+
+// alternativeImageSourcesIDMS returns unique list of DockerImageReference objects from list of ImageDigestMirrorSet objects
+// addSourceAsLastAlternate decides whether the original imageRef is first or the last element in the result
+func alternativeImageSourcesIDMS(imageRef reference.DockerImageReference, idmsList []apicfgv1.ImageDigestMirrorSet, addSourceAsLastAlternate bool) ([]reference.DockerImageReference, error) {
```
Comment (ardaguclu) [inline pkg/cli/image/strategy/onerror.go:221]: Apart from this function, the other changes are mostly copying of ICSP and renaming it to IDMS. But this function, we need to assure that it works properly. @QiWang19 could you please take a look at especially this function?.

File: pkg/cli/image/strategy/onerror.go (hunk 1 above)
Comment (QiWang19) [reply pkg/cli/image/strategy/onerror.go:229]: [mirrorSourcePolicy](https://github.com/openshift/api/blob/f36cfb9a990998106c5ebaa4fad0c7ff1264a2f5/config/v1/0000_10_config-operator_01_imagedigestmirrorset.crd.yaml#L50) defines whether crio pulls the image from the source registry when pulling. I think
-  if the mirrorSourcePolicy is specified as `NeverContactSource`, the source should not add to `imageSources`, neither the first time nor the last item, so the oc behavior is compatible with the container runtime. 
- if `NeverContactSource`is not `NeverContactSource` we use the same logic as current `alternativeImageSources()`

WDYT?

Comment (ardaguclu) [reply pkg/cli/image/strategy/onerror.go:229]: Sounds fine to me. But `idms.Spec.ImageDigestMirrors` is an array, what should be the behavior when one item in this array contains `NeverContactSource` and the others have `AllowContactingSource`?

**Comment by kasturinarra** [comment]
@ardaguclu I have finished pre-merge testing for this feature and adding the details of what has been tested

1. Latest oc image extract -h contains the --idms-file flag and does not contain --icsp-file flag
2.  When using --icsp-file flag with latest oc it throws a depercated message and suggests to use --idms-file but still continues to work for 4.14 release
3.  Converted icsp file to imagedigestmirror set file using command `oc adm migrate icsp <path_to_icsp_file>` removed source which is quay.io info from the imagedigestmirrorset file , tried to extract oc-mirror image using --idms-file flag and it works fine
4.  Created a new imagedigestmirrorset file with `mirrorSourcePolicy` set to `NeverContactSource`and tried to extract `oc-mirror` image with --idms-file flag and that works fine

Above is what i have tested, i will draft cases tomorrow morning and send it for review, can you let me know if i am missing something or any other scenario that needs to be covered ? thanks !!

File: pkg/cli/image/strategy/onerror.go (hunk 1 above)
Comment (QiWang19) [reply pkg/cli/image/strategy/onerror.go:229]: I think we can break the loop and error out. 
`NeverContactSource` and `AllowContactingSource` for the same source is conflicting. MCO returns the error when rolling out the IDMS CR.

**Comment by ardaguclu** [comment]
> 

Thanks @kasturinarra. These ^ test cases look good to me.

File: pkg/cli/image/strategy/onerror.go (hunk 1 above)
Comment (ardaguclu) [reply pkg/cli/image/strategy/onerror.go:229]: Thanks. I'll change that way.

Comment (ardaguclu) [reply pkg/cli/image/strategy/onerror.go:237]: @QiWang19 updated. Could you PTAL?

**Comment by ardaguclu** [comment]
@QiWang19 could you please take a look at this PR?. Thanks.

File: pkg/cli/image/strategy/onerror.go [hunk 3]
```diff
@@ -162,3 +154,180 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
...
+	var imageSources []reference.DockerImageReference
+	addSource, err := isAddSource(idmsList)
+	if err != nil {
+		return nil, err
+	}
+	if addSource {
+		klog.V(5).Infof("%v ImageReference added to potential ImageSourcePrefixes from ImageDigestMirrorSet", imageRef.AsRepository().AsV2())
+		if !addSourceAsLastAlternate {
+			imageSources = append(imageSources, imageRef.AsRepository().AsV2())
+		}
```
Comment (QiWang19) [inline pkg/cli/image/strategy/onerror.go:231]: like the other comment pointed out, the `isAddSource(idmsList, rdm.Source)` can be called inside the loop and 
adjust the `imageSources` after the loop.

File: pkg/cli/image/strategy/onerror.go [hunk 4]
```diff
@@ -162,3 +154,180 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
...
+				suffix = repo[len(rdm.Source):]
+			}
+			klog.V(5).Infof("%v ImageDigestMirrors source matches given image", imageRef.AsRepository().AsV2())
+			for _, m := range rdm.Mirrors {
+				mRef, err := reference.Parse(string(m) + suffix)
+				if err != nil {
+					return nil, fmt.Errorf("invalid mirror %q: %w", m, err)
+				}
+				imageSources = append(imageSources, mRef)
+				klog.V(5).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
```
Comment (QiWang19) [inline pkg/cli/image/strategy/onerror.go:260]:
```suggestion
				klog.V(5).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageDigestMirrorSet", m)
```

File: pkg/cli/image/strategy/onerror.go [hunk 5]
```diff
@@ -162,3 +154,180 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
...
+			}
+			// AsV2 in the right call is required to ensure we transform docker registry
+			// from docker.io to registry-1.docker.io
+			if imageRef.AsRepository().AsV2() != rdmSourceRef.AsRepository().AsV2() {
+				if !isSubrepo(repo, rdm.Source) {
+					continue
+				}
+				suffix = repo[len(rdm.Source):]
+			}
+			klog.V(5).Infof("%v ImageDigestMirrors source matches given image", imageRef.AsRepository().AsV2())
```
Comment (QiWang19) [inline pkg/cli/image/strategy/onerror.go:253]: we can call `isAddSource(idmsList, rdm.Source)` after the above check, so the `rdm.Source` is the registry source matches `imageRef`.

**Review by QiWang19 — COMMENTED** [review]
Thanks. @ardaguclu  I suggested changing to the `isAddSource` function based on the `mirrorSourcePolicy` usage.

File: pkg/cli/image/strategy/onerror.go (hunk 2 above)
Comment (QiWang19) [reply pkg/cli/image/strategy/onerror.go:221]: Sure, I paste the function below.
```go
func alternativeImageSourcesIDMS(imageRef reference.DockerImageReference, idmsList []apicfgv1.ImageDigestMirrorSet, addSourceAsLastAlternate bool) ([]reference.DockerImageReference, error) {
	var imageSources []reference.DockerImageReference
	var addSource bool
	repo := imageRef.AsRepository().Exact()
	for _, idms := range idmsList {
		repoDigestMirrors := idms.Spec.ImageDigestMirrors
		for _, rdm := range repoDigestMirrors {
			...
                        ...
			if imageRef.AsRepository().AsV2() != rdmSourceRef.AsRepository().AsV2() {
				if !isSubrepo(repo, rdm.Source) {
					continue
				}
				suffix = repo[len(rdm.Source):]
			}
			klog.V(5).Infof("%v ImageDigestMirrors source matches given image", imageRef.AsRepository().AsV2())
			// check valid mirrorSourcePolicy
			addSource, err = isAddSource(idmsList, rdm.Source)
			if err != nil {
				return nil, err
			}
			for _, m := range rdm.Mirrors {
				mRef, err := reference.Parse(string(m) + suffix)
				if err != nil {
					return nil, fmt.Errorf("invalid mirror %q: %w", m, err)
				}
				imageSources = append(imageSources, mRef)
				klog.V(5).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
			}
		}
	}
	if addSource && !addSourceAsLastAlternate {
		// prepend imageRef.AsRepository().AsV2()  to the  imageSources
	}
	if addSource && addSourceAsLastAlternate {
		// append imageRef.AsRepository().AsV2() to the imageSources
	}
	uniqueMirrors := make([]reference.DockerImageReference, 0, len(imageSources))
	uniqueMap := make(map[reference.DockerImageReference]bool)
	//  omit the rest of the function ...
	return uniqueMirrors, nil
}
```

File: pkg/cli/image/strategy/onerror.go [hunk 6]
```diff
@@ -162,3 +154,180 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
...
+	var mirrorSourcePolicy apicfgv1.MirrorSourcePolicy
+	for _, idms := range idmsList {
+		for _, mirror := range idms.Spec.ImageDigestMirrors {
+			if !found && mirror.MirrorSourcePolicy != "" {
+				mirrorSourcePolicy = mirror.MirrorSourcePolicy
+				found = true
+				continue
+			}
+			if mirrorSourcePolicy != mirror.MirrorSourcePolicy {
+				return found, fmt.Errorf("ImageDigestMirrorSet can only contain one MirrorSourcePolicy")
```
[the remark this answers is not in the document]
Comment (QiWang19) [reply pkg/cli/image/strategy/onerror.go:290]: Sure, I paste the function below.
```go
func alternativeImageSourcesIDMS(imageRef reference.DockerImageReference, idmsList []apicfgv1.ImageDigestMirrorSet, addSourceAsLastAlternate bool) ([]reference.DockerImageReference, error) {
	var imageSources []reference.DockerImageReference
	var addSource bool
	repo := imageRef.AsRepository().Exact()
	for _, idms := range idmsList {
		repoDigestMirrors := idms.Spec.ImageDigestMirrors
		for _, rdm := range repoDigestMirrors {
			...
                        ...
			if imageRef.AsRepository().AsV2() != rdmSourceRef.AsRepository().AsV2() {
				if !isSubrepo(repo, rdm.Source) {
					continue
				}
				suffix = repo[len(rdm.Source):]
			}
			klog.V(5).Infof("%v ImageDigestMirrors source matches given image", imageRef.AsRepository().AsV2())
			// check valid mirrorSourcePolicy
			addSource, err = isAddSource(idmsList, rdm.Source)
			if err != nil {
				return nil, err
			}
			for _, m := range rdm.Mirrors {
				mRef, err := reference.Parse(string(m) + suffix)
				if err != nil {
					return nil, fmt.Errorf("invalid mirror %q: %w", m, err)
				}
				imageSources = append(imageSources, mRef)
				klog.V(5).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
			}
		}
	}
	if addSource && !addSourceAsLastAlternate {
		// prepend imageRef.AsRepository().AsV2()  to the  imageSources
	}
	if addSource && addSourceAsLastAlternate {
		// append imageRef.AsRepository().AsV2() to the imageSources
	}
	uniqueMirrors := make([]reference.DockerImageReference, 0, len(imageSources))
	uniqueMap := make(map[reference.DockerImageReference]bool)
	//  omit the rest of the function ...
	return uniqueMirrors, nil
}
```

**Comment by ardaguclu** [comment]
@QiWang19 could you PTAL for one last review?. Thanks.

File: pkg/cli/image/strategy/onerror.go [hunk 7]
```diff
@@ -162,3 +154,185 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
...
+	}
+	if len(alternates) == 0 {
+		return nil, fmt.Errorf("no alternative image references found for image: %s", locator.String())
+	}
+	s.alternates[locator] = alternates
+	return s.alternates[locator], nil
+}
+
+// resolve gathers possible image sources for a given image
+// gathered from ImageContentSourcePolicy file.
```
Comment (QiWang19) [inline pkg/cli/image/strategy/onerror.go:200]:
```suggestion
// gathered from ImageDigestMirrorSet file.
```

File: pkg/cli/image/strategy/onerror.go [hunk 8]
```diff
@@ -162,3 +154,185 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
...
+				continue
+			}
+			if !found && mirror.MirrorSourcePolicy != "" {
+				mirrorSourcePolicy = mirror.MirrorSourcePolicy
+				found = true
+				continue
+			}
+			if mirrorSourcePolicy != mirror.MirrorSourcePolicy {
+				return found, fmt.Errorf("ImageDigestMirrorSet can only contain one MirrorSourcePolicy for source %s", source)
+			}
```
Comment (QiWang19) [inline pkg/cli/image/strategy/onerror.go:296]: AllowContactingSource is the default value, empty mirror.MirrorSourcePolicy and AllowContactingSource are not equal but valid. eliminate that from the error return.

```suggestion
			if !found {
                if mirror.MirrorSourcePolicy != "" {
                    mirrorSourcePolicy = mirror.MirrorSourcePolicy
                } else {
                    mirrorSourcePolicy = apicfgv1.AllowContactingSource
                }
				found = true
				continue
			}
            if mirrorSourcePolicy == apicfgv1.AllowContactingSource && mirror.MirrorSourcePolicy == "" {
                continue
            }
            if mirrorSourcePolicy != mirror.MirrorSourcePolicy {
				return found, fmt.Errorf("ImageDigestMirrorSet can only contain one MirrorSourcePolicy for source %s", source)
			}
```

Comment (ardaguclu) [reply pkg/cli/image/strategy/onerror.go]: @QiWang19 updated, PTAL?

**Review by QiWang19 — COMMENTED** [review]
/lgtm

**Pull request merged**